File: merge_sort/merge_sort.java
public class merge_sort {
    static int[] tempArr = new int[10];

    public static void mergeSort(int[] arr, int first, int last) {
        if (first == last)
            return;
        int mid = (first + last) >> 1;
        mergeSort(arr, first, mid);
        mergeSort(arr, mid + 1, last);

        int left = first;
        int right = mid + 1;
        int limit = 0;
        while (left <= mid || right <= last) {
            if (left > mid)
                tempArr[limit++] = arr[right++];
            else if (right > last)
                tempArr[limit++] = arr[left++];
            else if (arr[left] < arr[right])
                tempArr[limit++] = arr[left++];
            else
                tempArr[limit++] = arr[right++];
        }

        for (int i = 0; i < limit; ++i)
            arr[first + i] = tempArr[i];
    }

    public static void main(String[] args) {
        int[] arr = { 2, 7, 8, 4, -9, 0, -100, 20, 45, 3 };
        mergeSort(arr, 0, 9);
        for (int index = 0; index < 10; ++index)
            System.out.printf("%d ", arr[index]);
    }
}

File: merge_sort/merge_sort.py
tempArr = [0] * 10

def mergeSort(arr, first, last):
    if first == last:
        return
    mid = (first + last) >> 1
    mergeSort(arr, first, mid)
    mergeSort(arr, mid + 1, last)

    left = first
    right = mid + 1
    limit = 0
    while (left <= mid) or (right <= last):
        if left > mid:
            tempArr[limit] = arr[right]
            limit += 1
            right += 1
        elif right > last:
            tempArr[limit] = arr[left]
            limit += 1
            left += 1
        elif arr[left] < arr[right]:
            tempArr[limit] = arr[left]
            limit += 1
            left += 1
        else:
            tempArr[limit] = arr[right]
            limit += 1
            right += 1

    i = 0
    while i < limit:
        arr[first + i] = tempArr[i]
        i += 1


arr = [2, 7, 8, 4, -9, 0, -100, 20, 45, 3]
mergeSort(arr, 0, 9)
print(arr)
